.design/requirements-575.md

```diff
@@ -0,0 +1,269 @@
+
+HTTP server implementations can provide support for WebSocket, but their configuration is complex in the context of the OSGi Whiteboard specification for Jakarta Servlet.
+
+A Jakarta Websocket Whiteboard would ease the deployment of WebSocket handlers in an OSGi HTTP environment.
+This document describes the requirements for such a specification.
+
+## Terminology
+
+* *WebSocket* - The communication channel based on the WebSocket protocol.
+* *WebSocket frame* - Fragment of a WebSocket message as transferred on the WebSocket.
+* *WebSocket message* - Data or control message transferred over the WebSocket. A message can be transmitted via one or multiple frames. A data message can be text or binary. A control message can be a ping, pong or close message.
+* *WebSocket Container* - The server-side, implementation-specific, container of server WebSocket Endpoints.
+* *WebSocket Session* - Description of a single client-server WebSocket connection.
+* *WebSocket Endpoint* - Handler of a WebSocket session lifecycle: open, close, error. The endpoint registers frame listeners when the session is opened. Can be used either on client or server side.
+* *WebSocket (Message) Handler* - Handler notified of WebSocket message reception. Can be used either on client or server side.
+
+## Problem Description
+
+Registering a WebSocket handler can currently be done by using a custom configuration of an HTTP servlet.
+The problem is that such configuration depends on the underlying HTTP server implementation.
+
+For example, the Eclipse Jetty server requires registering components before initializing an HTTP servlet to allow it to handle the WebSocket protocol.
+This requires to write code specific to Jetty in an OSGi HTTP whiteboard environment.
+There is also an issue between the Jetty WebSocket server components and the servlet context given to it by the Apache Felix HTTP Jetty implementation.
+
+As a result, the registration of WebSocket in the current state of the specification goes against the loose coupling concept usually provided by OSGi.
+
+What is needed is a WebSocket whiteboard specification moving specific configuration to the HTTP/WebSocket whiteboard implementation.
+
+Reusing parts, if not all, of the Jakarta WebSocket specification will ease the development of WebSocket handlers and avoid redefining existing APIs.
+It would also allow an easy reuse of existing Jakarta EE WebSocket handlers.
+
+## Note on *Partial* vs. *Whole* messages handlers
+
+The Jakarta WebSocket specification allows two kinds of message handlers which are described as:
+* *Whole* message handlers are notified by the container on arrival of a complete message (all frames recombined).
+* *Partial* message handlers are notified by the implementation when it is ready to deliver frames of a whole message.
+
+As it is explicitly stated in the JavaDoc part of the specification, the *partial* message handler is implementation-dependent.
```

Comment:
This is unclear, the difference between partial and whole is if the message is complete or not, while handling partly messages might be dependent on the protocol implementation it does not mean it is specific to the Webservice API implementation.